codegen/smithy-aws-swift-codegen/src/main/kotlin/software/amazon/smithy/aws/swift/codegen/middleware/OperationEndpointResolverMiddleware.kt

```diff
@@ -0,0 +1,78 @@
+import software.amazon.smithy.swift.codegen.middleware.MiddlewareStep
+import software.amazon.smithy.swift.codegen.model.getTrait
+import software.amazon.smithy.swift.codegen.utils.toCamelCase
+
+/**
+ * Generates EndpointResolverMiddleware interception code.
+ * Including creation of EndpointParams instance and pass it as middleware param along with EndpointResolver
+ */
+class OperationEndpointResolverMiddleware(
+    val ctx: ProtocolGenerator.GenerationContext,
+) : MiddlewareRenderable {
+
+    override val name = "EndpointResolverMiddleware"
+
+    override val middlewareStep = MiddlewareStep.BUILDSTEP
+
+    override val position = MiddlewarePosition.BEFORE
+
+    override fun render(writer: SwiftWriter, op: OperationShape, operationStackName: String) {
+        val output = MiddlewareShapeUtils.outputSymbol(ctx.symbolProvider, ctx.model, op)
+        val outputError = MiddlewareShapeUtils.outputErrorSymbol(op)
+        val params = mutableListOf<String>()
+        ctx.service.getTrait<EndpointRuleSetTrait>()?.ruleSet?.let { node ->
+            val ruleSet = EndpointRuleset.fromNode(node)
+            val staticContextParams = op.getTrait<StaticContextParamsTrait>()?.parameters ?: emptyMap()
+            val clientContextParams =
+                ctx.service.getTrait<ClientContextParamsTrait>()?.parameters ?: emptyMap()
+            val parameters = ruleSet.parameters.toList()
+            parameters.toList()
+                .sortedBy { it.name.toString() }
+                .forEach { param ->
+                    val memberName = param.name.toString().toCamelCase()
+                    val contextParam = ctx.model.expectShape(op.inputShape).members()
+                        .firstOrNull { it.getTrait<ContextParamTrait>()?.name == param.name.toString() }
+                    if (staticContextParams.containsKey(param.name.toString())) {
+                        // check static context params
+                        params.add("$memberName: ${staticContextParams[param.name.toString()]?.value}")
+                    } else if (contextParam != null) {
+                        // check context params that comes from the input shape
+                        params.add("$memberName: input.${contextParam.memberName.toCamelCase()}")
+                    } else if (param.builtIn.getOrNull() != null) {
+                        // check the built-in params that comes from the client config
+                        val defaultValue = if (param.isRequired) " ?? ${param.defaultValue.getOrNull()}" else ""
+                        params.add("$memberName: config.${param.name.toString().toCamelCase()}$defaultValue")
+                    } else if (param.defaultValue.getOrNull() != null) {
+                        // check if the param has a default value
+                        params.add("$memberName: ${param.defaultValue.getOrNull().toString().toCamelCase()}")
+                    }
```

Comment:
**Question**: When might none of these conditions are true? Right now you'd omit that parameter...is that right?

Reply:
Added a clarification.